cloudify_azure/resources/deployment.py

```diff
@@ -163,3 +148,130 @@ def delete(ctx, **_):
+    properties.update(kwargs)
+    params = format_params(properties.get('params', {}))
+    return properties, params
+
+
+def get_resource_group_name_deployment_name_and_api_version(ctx):
+    deployment_name = utils.get_resource_name(ctx)
+    resource_group_name = ctx.node.properties.get(
+        'resource_group_name', deployment_name)
+    api_version = \
+        ctx.node.properties.get('api_version', constants.API_VER_RESOURCES)
+    return deployment_name, resource_group_name, api_version
+
+
+def execute_what_if(deployment,
+                    resource_group_name,
+                    deployment_name,
+                    template,
+                    params):
+    what_if_properties = {
+        'mode': DeploymentMode.incremental,
+        'template': template,
+        'parameters': params
+    }
+    try:
+        return deployment.what_if(resource_group_name,
+                                  deployment_name,
+                                  what_if_properties)
+    except CloudError:
+        raise cfy_exc.NonRecoverableError(
+            "What if operation on deployment {dep_name} failed. Can't "
+            "calculate accurate state.".format(dep_name=deployment_name))
+
+
+def calculate_state(ctx, initial_resources, actual_resources, what_if_res):
+    """
+    Create a list of live resources of the deployment.
+    Save this list in the state runtime property.
+    :param initial_resources: list of resources id`s created by the deployment.
+    :param actual_resources: list of resources id`s that exists in resource
+    group.
+    :param: result of what-if operation in order to indicate resources that
+    not appear in actual_resources.
+    """
+    state = []
+    initial_ids = [resource['id'] for resource in initial_resources]
+    actual_ids = [resource['id'] for resource in actual_resources]
+    for resource_id in initial_ids:
+        if resource_id not in actual_ids and not\
+                check_if_resource_alive_in_what_if_result(resource_id,
+                                                          what_if_res):
+            ctx.logger.debug("Resource {resource} not exists, deployment is "
+                             "drifted.".format(resource=resource_id))
+            continue
+        state.append(resource_id)
+
+    ctx.instance.runtime_properties[STATE] = state
+    ctx.instance.runtime_properties[IS_DRIFTED] = \
+        False if state == initial_ids else True
+
+
+def check_if_resource_alive_in_what_if_result(resource_id, what_if_result):
+    """
+    Given resource id and what-if operation result,
+    check if the resource is alive using the what-if operation result.
+    The change types of the what-if operation described here:
+
+    https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/
+    template-deploy-what-if?tabs=azure-powershell#change-types
+
+    """
+    status = what_if_result.get('status')
+    if status != 'Succeeded':
```

Comment:
Are there any other valid statuses?

Reply:
As you can see their docs are awesome:
https://docs.microsoft.com/en-us/python/api/azure-mgmt-resource/azure.mgmt.resource.resources.v2019_10_01.models.whatifoperationresult?view=azure-python
But I found this:
https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/async-operations
so I guess only Succeeded is valid(Canceled/Failed are not valid).